### ISSUE TITLE
[no-unused-vars] Prevent false positives with classes 

### ISSUE THREAD

**Issue by luxaritas** [typescript-eslint/typescript-eslint#586]
**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/no-unused-vars": ["error"]
  }
}
```

```TS
class Base {
    public f(value) {
        throw new Error("IllegalOperationError");
    }
}

class Impl extends Base {
    public f(value) {
        console.log(value);
    }
}

class SecondaryImpl extends Base {
    public f(value) {
        console.log("Irrelevant");
    }
}
```

**Expected Result**
No violation

**Actual Result**
Violation (twice)

**Additional Info**
For some reason when I filed #52, I apparently thought this rule already did this, not sure why, but it's the same idea.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `1.9.0` |
| `@typescript-eslint/parser`        | `1.9.0` |
| `TypeScript`                       | `3.3.4000` |
| `ESLint`                           | `5.15.3` |
| `node`                             | `8.11.3` |
| `npm`                              | `6.9.0` |

**Comment by bradzacher** [comment]
you mention there's two violations.
What are they? What lines?

**Comment by luxaritas** [comment]
Whoops sorry.
`'value' is defined but never used (@typescript-eslint/no-unused-vars)`, on both lines 2 and 14.

**Comment by bradzacher** [comment]
to me those violations look correct.
I suspect you missed something in your example code, because currently you have 3 isolated classes, 2 of which have unused parameters.

**Comment by luxaritas** [comment]
Yikes, the latter two classes were supposed to extend Base. Fixed in the example code.

**Comment by bradzacher** [comment]
Revisiting this - this will never be a truly supported use case.
Detecting that a class has a subclass that defines a method which uses the argument is very complex and (unless I'm mistaken) it's not possible with typescript's current checker API.

You can get around this using the following rule config, and by deleting the unused argument in `SecondaryImpl`. This is both valid TypeScript code, and lints fine.

```ts
/* eslint no-unused-vars: ['error', { argsIgnorePattern: "^_" }] */
class Base {
	public f(_value: string) {
        throw new Error("IllegalOperationError");
    }
}

export class Impl extends Base {
    public f(value: string) {
        console.log(value);
    }
}

export class SecondaryImpl extends Base {
    public f() {
        console.log("Irrelevant");
    }
}
```

[typescript repl](https://www.typescriptlang.org/play/#code/MYGwhgzhAEBCkFNoG8BQBIADgVwEYgEthoAzACgH0A3MEbBALmggBcAnAgOwHMBKFVNCHChLABZsA9gHdonBLICibKWzIAiAJIgQCbrQDymBGzAsCkzstXreAbkHCAvqheoEAD0yS2LaKEgYTQBbTBBoTxYETgATGHgIJDQRHHwiUjIaOkZmdi4+ARERYEsISV0AOhBJbkzaentHIRc3T29ff3AoaABlBBLYsDYATxCwiI8o2PjEQuFUwmJyfmSioQGyyurarRUEXRpOFlsHERagA)
[eslint repl](https://eslint.org/demo/#eyJ0ZXh0IjoiLyogZXNsaW50IG5vLXVudXNlZC12YXJzOiBbJ2Vycm9yJywgeyBhcmdzSWdub3JlUGF0dGVybjogXCJeX1wiIH1dICovXG5jbGFzcyBCYXNlIHtcblx0ZihfdmFsdWUpIHtcbiAgICAgICAgdGhyb3cgbmV3IEVycm9yKFwiSWxsZWdhbE9wZXJhdGlvbkVycm9yXCIpO1xuICAgIH1cbn1cblxuZXhwb3J0IGNsYXNzIEltcGwgZXh0ZW5kcyBCYXNlIHtcbiAgICBmKHZhbHVlKSB7XG4gICAgICAgIGNvbnNvbGUubG9nKHZhbHVlKTtcbiAgICB9XG59XG5cbmV4cG9ydCBjbGFzcyBTZWNvbmRhcnlJbXBsIGV4dGVuZHMgQmFzZSB7XG4gICAgZigpIHtcbiAgICAgICAgY29uc29sZS5sb2coXCJJcnJlbGV2YW50XCIpO1xuICAgIH1cbn1cbiIsIm9wdGlvbnMiOnsicGFyc2VyT3B0aW9ucyI6eyJlY21hVmVyc2lvbiI6OSwic291cmNlVHlwZSI6Im1vZHVsZSIsImVjbWFGZWF0dXJlcyI6eyJqc3giOnRydWV9fSwicnVsZXMiOnt9LCJlbnYiOnsiY29tbW9uanMiOmZhbHNlLCJub2RlIjpmYWxzZX19fQ==)
